cubeview.1
.\" $Id: cubeview.1,v 1.2 2010-05-18 17:09:57 paumard Exp $
.\"
.IX Title "CUBEVIEW 1"
.TH CUBEVIEW 1 "2008\-01\-03" "" "Yorick"
.SH "NAME"
cubeview \- view 3D FITS files
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBcubeview\fR [\fB\-\-pixel\fR[=\fI0/1\fR]] [\fB\-\-stand\-alone\fR[=\fI0/1\fR]] [\fB\-\-ui\fR=\fIui\fR] [\fIfile\fR]
.br
or
\&\fByorick \-i cubeview.i\fR [options] [\fIfile\fR]
.br
or from within \fIyorick\fR(1): \fBcubeview\fR,options or \fBcv_gtk\fR[,options].
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Cubeview is a 3D data viewer specialized in spectro\-imaging,
implemented using the Yorick interpreted language (see
\fIyorick\fR(1)). A 3D data cube in the sense of cubeview is a
three\-dimensional array of numbers, usually stored in a FITS file.
.PP
Cubeview can function either as a stand\-alone viewer for viewing 3D
FITS files (in which case you don't need to know much about Yorick to
use it) or as a Yorick package, in which case it is possible to view
Yorick arrays directly. Cubeview can be customized and enhanced
through "hooks" which can automatically perform custom actions while
the cube is being viewed. For instance, it is possible to overplot
some complex contour map over the slice view each time it is
refreshed. For more details about the cubeview API in Yorick, read
cubeview.i. When cubeview is launched from a terminal window, it is
possible to type Yorick commands in that window almost at any time. In
the following, this manpage assumes the reader is not a regular Yorick
user.
.PP
Since Cubeview is specialized in spectro\-imaging, it assumes the first
two dimensions of the cube are of spatial nature while the third is
spectral. Cubeview is able to correclty interpret the FITS headers of
data produced with the decommissioned BEAR instrument which used to
be operated at CFHT and SINFONI currently operated at ESO VLT. Other
data may be interpreted if they follow the same conventions. If the
FITS header cannot be interpreted, the axes in the plots cannot be
trusted, but you can still explore the 3D cube.
.PP
Cubeview uses three windows: a toolbox, a slice image window and a
spectrum plot window. The toolbox allows to open a new FITS file, save
the currently selected sub\-cube (determined both by the spectrum and
slice being viewed), set various display parameters, and perform
various actions, most notably selecting a new spectrum or a new
slice. If \fIfile\fR is set in the calling sequence, then all three
windows open at once, else only the toolbox appears at first, allowing
the user to select a file to read.
.SH "MAIN PAGE"
The \fIMain\fR page in the toolbox offers reasonably self\-explanatory
buttons to perform various actions. To select a new slice, click on
\fISlice\fR in the \fISelect\fR frame, then drag the mouse pointer
over the region of interest in the spectrum window. Conversely, a new
sepctrum is selected by first clicking \fISpectrum\fR and then using
the mouse in the slice window. How you use the mouse for selecting a
spectrum depends on the \fIAperture type\fR selected in the
\fISpectrum\fR property page:
.IP \fICircular\fR 4
left button: click to select new center; right button: drag from new
center to new edge;
.IP \fISquare\fR 4
left button: click to select new center; right button: drag from new
center to new edge;
.IP \fIRectangular\fR 4
drag from one corner to the opposite.
.SH "SPECTRUM PAGE"
The \fISpectrum\fR property page allows to select the \fIAperture
type\fR mentioned above. In addition, if the FITS header has been
interpreted correctly, it is possible to switch the spectral axis
between \fIWavelength\fR, \fIFrequency\fR, \fIChannels\fR (raw indices
in the cube, the only meaningful value if the header was not
interpreted correctly) and \fIVelocity\fR relative to a \fIReference
wavelength\fR which can also be set on this page. \fISmoothing FWHM\fR
controls whether the displayed spectrum should be Gaussian\-smoothed to
increase the apparent signal\-to\-noise.
.SH "SLICE PAGE"
The slice can be displayed in two modes. The most usual one (named
\fINormal (palette)\fR in Cubeview) is palette\-based. The
corresponding \fIColor palette\fR can be selected among the standard
Yorick ones. Alternatively, Cubeview can produce three\-color images
using virtual, overlapping red, green and blue filters. The slice can
then be displayed either at 8 or 24 bit color\-depth. 24 bit color
depth is usually better, but 8 bit is useful to save to some image
formats, which you can do from the Yorick command line. \fISmoothing
FWHM\fR and \fIOversampling\fR control two means of smoothing the
displayed image for eye candy.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-pixel\fR[=(false|true)]" 4
Control whether the slice window axes are in pixels (true) or in world
coordinates as indicated by the CDELT, CRPIX and CRVAL FITS
cards. \-\-pixel is equivalent to \-\-pixel=true. The default is true.
.IP "\fB\-\-stand\-alone\fR[=(false|true)]" 4
Control whether closing the toolbox window exits Yorick. \-\-stand\-alone
is equivalent to \-\-stand\-alone=true. This is the default for he first
form of invocation. If set to \fIfalse\fR, it is necessary to type
"quit" at the Yorick prompt to completely quit the application.
.IP "\fB\-\-ui\fR=(gtk|tws|text)" 4
Control the look\-and\-feel of the toolbox. The default toolbox uses the
GTK toolkit, and requires several software components in addition to
Yorick (python, pygtk and libglade). An alternative toolbox coded
entirely in Yorick is also available. It uses the "TWS" package to
draw buttons and other widgets. It is uglier, but more portable, than
the GTK\-based toolbox. Finally, it is possible to completely control
cubeview from the Yorick prompt, in "text" mode. Type "cv_library" for
a list of available commands.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIyorick\fR\|(1), cubeview.i
.SH "AUTHORS"
.IX Header "AUTHORS"
Thibaut Paumard <paumard@users.sourceforge.net>


changelog
cubeview (1.4)

  * improve export capabilities: export window plot or data,
    e.g. slice to FITS...
  * add icons and .desktop file

 -- Thibaut Paumard <paumard@users.sourceforge.net>  Wed, 05 Mar 2008 09:46:52 +0100

cubeview (1.3)

  * add functions to export a view to an image file (eps, pdf, jpg or png)

 -- Thibaut Paumard <paumard@users.sourceforge.net>  Mon, 03 Mar 2008 21:25:46 +0100

cubeview (1.2.1)

  * clean files moved to yutils
  * update documentation

 -- Thibaut Paumard <paumard@users.sourceforge.net>  Thu, 17 Jan 2008 15:42:36 +0100

cubeview (1.2)

  * add files:
      * manpage cubeview.1
      * changelog
      * Makefile
      * cubeview_start.i
  * switch back from newfits.i to fits.i
  * split some items out of cv_interns:
    cv_interns.cube   -> cv_cube (performance)
              .valids -> cv_valids (performance)
              .header -> cv_fh (a list cannot be a struct member)

 -- Thibaut Paumard <paumard@users.sourceforge.net>  Thu,  3 Jan 2008 14:58:56 +0100

Local variables:
  mode: debian-changelog


cubeview.py
#!/usr/bin/env python
# cubeview.py
# $Id: cubeview.py,v 1.2 2010-04-07 20:26:43 paumard Exp $
# Inspired from yao.py
#     This file is part of Cubeview.
#     Copyright (C) 2007  Thibaut Paumard <paumard@users.sourceforge.net>
#
#     This program is free software; you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation; either version 2 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License along
#     with this program; if not, write to the Free Software Foundation, Inc.,
#     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import gtk
import gtk.glade
import sys
import gobject
import os, fcntl, errno
from time import *

class cubeview:
   
   def destroy(self, wdg, data=None):
      self.py2yo('cv_suspend')
      gtk.main_quit()
      
   def __init__(self,cubeview_glade):
      self.cubeview_glade = cubeview_glade
      self.usercmd = 'STOP'
      
      # callbacks and glade UI
      dic = {
         'cv_handler': self.cv_handler,
         'cv_palette_handler': self.cv_palette_handler,
         'cv_box_handler': self.cv_box_handler,
         'cv_sltype_handler': self.cv_sltype_handler,
         'cv_depth_handler': self.cv_depth_handler,
         'cv_set_sptype': self.cv_set_sptype,
         'cv_set_aperture': self.cv_set_aperture,
         'on_quitter1_activate': self.on_quitter1_activate,
         'on_ouvrir1_activate': self.on_ouvrir1_activate,
         'on_save_selection_as_activate': self.on_save_selection_as_activate,
         'cv_switch_to': self.cv_switch_to,
         'cv_about': self.cv_about,
         'cv_export_window': self.cv_export_window,
         'on_export_plot_or_data': self. on_export_plot_or_data,
         'set_export_formats_list': self.set_export_formats_list,
        }

      self.py2yo('cv_gtk_init')
      self.glade = gtk.glade.XML(self.cubeview_glade)
      self.window = self.glade.get_widget('toolbox')
      self.glade.get_widget('export-data1').set_active(1)
      self.glade.get_widget('export-slice1').set_active(1)
      self.set_export_formats_list(None)
      if (self.window):
         self.window.connect('destroy', self.destroy)
      self.glade.signal_autoconnect(dic)

      # set stdin non blocking, this will prevent readline to block
      fd = sys.stdin.fileno()
      flags = fcntl.fcntl(fd, fcntl.F_GETFL)
      fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
      
      # add stdin to the event loop (yorick input pipe by spawn)
      gobject.io_add_watch(sys.stdin,gobject.IO_IN|gobject.IO_HUP,self.yo2py,None)

      # run
      gtk.main()

   def cv_switch_to(self,wdg):
      if wdg.get_name()=="spectrum_properties":
         page=1
      elif wdg.get_name()=="slice_properties":
         page=2
      self.glade.get_widget("notebook1").set_current_page(page)

   def cv_about(self,wdg):
      dialog = self.glade.get_widget('aboutdialog')
      dialog.run()
      dialog.hide()

# File menu hanlders
   
   def on_quitter1_activate(self,wdg):
      self.py2yo('quit')

   def on_ouvrir1_activate(self,wdg):
      chooser = gtk.FileChooserDialog(title='Open 3D FITS File',action=gtk.FILE_CHOOSER_ACTION_OPEN,buttons=(gtk.STOCK_CANCEL,gtk.RESPONSE_CANCEL,gtk.STOCK_OPEN,gtk.RESPONSE_OK))
      filter = gtk.FileFilter()
      filter.add_pattern('*.[fF][iI][tT][sS]')
      filter.add_pattern('*.[fF][iI][tT]')
      filter.add_pattern('*.[fF][iI][tT][sS].gz')
      filter.add_pattern('*.[fF][iI][tT].gz')
      filter.set_name('FITS Files')
      chooser.add_filter(filter)
      res = chooser.run()
      if res == gtk.RESPONSE_OK:
         file=chooser.get_filename()
         self.py2yo('cv_init "'+file+'"')
         self.py2yo('cv_gtk_init')
      chooser.destroy()

   def on_save_selection_as_activate(self,wdg):
      chooser = gtk.FileChooserDialog(title='Save to 3D FITS File',action=gtk.FILE_CHOOSER_ACTION_SAVE,buttons=(gtk.STOCK_CANCEL,gtk.RESPONSE_CANCEL,gtk.STOCK_OPEN,gtk.RESPONSE_OK))
      filter = gtk.FileFilter()
      filter.add_pattern('*.[fF][iI][tT][sS]')
      filter.add_pattern('*.[fF][iI][tT]')
      filter.add_pattern('*.[fF][iI][tT][sS].gz')
      filter.add_pattern('*.[fF][iI][tT].gz')
      filter.set_name('FITS Files')
      chooser.add_filter(filter)
      res = chooser.run()
      if res == gtk.RESPONSE_OK:
         file=chooser.get_filename()
         self.py2yo('cv_save_sel "'+file+'"')
      chooser.destroy()

   ## Export dialog:
   # open export dialog
   def cv_export_window(self,wdg):
      self.py2yo('cv_freemouse')
      chooser = self.glade.get_widget('export-window')
      filter = gtk.FileFilter()
      filter.add_pattern('*')
      filter.set_name('All files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.[fF][iI][tT][sS]')
      filter.add_pattern('*.[fF][iI][tT]')
      filter.add_pattern('*.[fF][iI][tT][sS].gz')
      filter.add_pattern('*.[fF][iI][tT].gz')
      filter.set_name('FITS files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.txt')
      filter.add_pattern('*.dat')
      filter.add_pattern('*.csv')
      filter.set_name('Text files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.jpg')
      filter.add_pattern('*.jpeg')
      filter.add_pattern('*.jfif')
      filter.set_name('JPEG files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.png')
      filter.set_name('PNG files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.pnm')
      filter.add_pattern('*.ppm')
      filter.set_name('PNM files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.eps')
      filter.set_name('EPS files')
      chooser.add_filter(filter)
      filter = gtk.FileFilter()
      filter.add_pattern('*.pdf')
      filter.set_name('PDF files')
      chooser.add_filter(filter)
      res = chooser.run()
      if res:
         savedata = self.glade.get_widget('export-data1').get_active()
         if self.glade.get_widget('export-slice1').get_active():
            what="slice"
         else:
            what="spectrum";
         selection = self.glade.get_widget('export-selection1').get_active()
         format = self.glade.get_widget('export-format1').get_active_text()
         filename = chooser.get_filename()
         if filename:
            self.py2yo('cv_export_misc "'+filename+'" "'+format+'" "'+what+'" '+str(int(savedata))+' '+str(int(selection)))
         else:
            self.py2yo('cv_warning "No filename selected: file not saved"')
      chooser.hide()

   # Plot <-> Data toggle
   def on_export_plot_or_data(self,wdg):
      if self.glade.get_widget('export-data1').get_active():
         self.glade.get_widget('export-selection1').set_sensitive(1);
      else:
         self.glade.get_widget('export-selection1').set_sensitive(0);
      self.set_export_formats_list(wdg)

   def set_export_formats_list(self,wdg):
      if self.glade.get_widget('export-data1').get_active():
         model=gtk.ListStore(str)
         iter=model.append()
         model.set(iter,0,'FITS')
         if self.glade.get_widget('export-slice1').get_active():
            iter=model.append()
            model.set(iter,0,'PNM')
            iter=model.append()
            model.set(iter,0,'PNG')
            iter=model.append()
            model.set(iter,0,'JPEG')
         else:
            iter=model.append()
            model.set(iter,0,'ASCII')
      else:
         model=gtk.ListStore(str)
         iter=model.append()
         model.set(iter,0,'EPS')
         iter=model.append()
         model.set(iter,0,'PDF')
         iter=model.append()
         model.set(iter,0,'PNG')
         iter=model.append()
         model.set(iter,0,'JPEG')
      self.glade.get_widget('export-format1').set_model(model=model)
      self.glade.get_widget('export-format1').set_active(0)
      
   def warning(self,msg):
      mbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK, 'Cubeview Warning');
      mbox.format_secondary_markup(msg);
      #,message-type=gtk.MESSAGE_WARNING,buttons=gtk.BUTTONS_OK);
      res=mbox.run();
      mbox.destroy();

# toolbox handlers

   def  cv_handler(self,wdg):
      self.py2yo('cv_freemouse')
      sleep(0.5)
      self.py2yo(wdg.get_name())
         
   def  cv_palette_handler(self,wdg):
      child=wdg.child
      child=wdg.get_child()
      self.py2yo('cv_set_palette "%s"' % child.get_text())
      self.glade.get_widget('Normal').set_active(1)

   def  cv_box_handler(self,wdg):
      self.py2yo('cv_set_' + wdg.get_name() + ' ' + wdg.get_text())
      
   def  cv_sltype_handler(self,wdg):
      if wdg.get_active():
         self.py2yo('cv_sltype "%s"' % wdg.get_name())
      
   def  cv_depth_handler(self,wdg):
      if wdg.get_active():
         self.py2yo('cv_depth "%s"' % wdg.get_name())
         self.glade.get_widget('3 color').set_active(1)
   
      
   def  cv_set_sptype(self,wdg):
      if wdg.get_active():
         self.py2yo('cv_set_sptype "%s"' % wdg.get_name())
      
   def  cv_set_aperture(self,wdg):
      if wdg.get_active():
         self.py2yo('cv_set_aperture "%s"' % wdg.get_name())

   def cv_init(self,sptype,aperture,spsmooth,refwl,
               sltype,slpalette,sldepth,slsmooth,overs):
      self.glade.get_widget(sptype).set_active(1);
      self.glade.get_widget(aperture).set_active(1);
      self.glade.get_widget('refwl').set_text(refwl);
      self.glade.get_widget('spsmooth').set_text(spsmooth);
      entry=self.glade.get_widget('slpalette').child;
      entry.set_text(slpalette);
      self.glade.get_widget(sldepth).set_active(1);
      self.glade.get_widget('slsmooth').set_text(slsmooth);
      self.glade.get_widget('overs').set_text(overs);
      self.glade.get_widget(sltype).set_active(1);
      self.py2yo('cv_graphicwindows 1')
     
   #
   # Yorick to Python Wrapper Functions
   #

   def y_parm_update(self,name,val):
      self.glade.get_widget(name).set_value(val)

   def y_text_parm_update(self,name,txt):
      self.glade.get_widget(name).set_text(txt)

   def y_set_checkbutton(self,name,val):
      self.glade.get_widget(name).set_active(val)
      
   def pyk_error(self,msg):
      dialog = gtk.MessageDialog(type=gtk.MESSAGE_ERROR,buttons=gtk.BUTTONS_OK,message_format=msg)
      dialog.run()
      dialog.destroy()
      
   def pyk_info(self,msg):
      dialog = gtk.MessageDialog(type=gtk.MESSAGE_INFO,buttons=gtk.BUTTONS_OK,message_format=msg)
      dialog.run()
      dialog.destroy()

   def pyk_info_w_markup(self,msg):
      dialog = gtk.MessageDialog(type=gtk.MESSAGE_INFO,buttons=gtk.BUTTONS_OK)
      dialog.set_markup(msg)
#      dialog.set_size_request(600,-1)
      dialog.run()
      dialog.destroy()

   def pyk_warning(self,msg):
      dialog = gtk.MessageDialog(type=gtk.MESSAGE_WARNING,buttons=gtk.BUTTONS_OK,message_format=msg)
      dialog.run()
      dialog.destroy()
      
   def pyk(self,msg):
      # sends string command to yorick
      sys.stdout.write(msg)
      sys.stdout.flush()
      
      # if this flag set, yorick is blocked waiting for tyk_resume
      _tyk_blocked=0
      
   def pyk_sync(self):
      _tyk_blocked=1
      sys.stdout.write('-s+y-n+c-+p-y+k-')
      sys.stdout.flush()
      
   def pyk_resume(self,msg):
      sys.stdout.write('pyk_resume'+msg)
      sys.stdout.flush()
      _tyk_blocked=0
   
   #
   # minimal wrapper for yorick/python communication
   #

   def yo2py_flush(self):
      sys.stdin.flush()
   
   def py2yo(self,msg):
      # sends string command to yorick's eval
      sys.stdout.write(msg+'\n')
      sys.stdout.flush()
   
   def yo2py(self,cb_condition,*args):
      if cb_condition == gobject.IO_HUP:
         raise SystemExit, "lost pipe to yorick"
      # handles string command from yorick
      # note: inidividual message needs to end with /n for proper ungarbling
      while 1:
         try:
            msg = sys.stdin.readline()
            msg = "self."+msg
            exec(msg)
            self.pyk_resume(msg)
         except IOError, e:
            if e.errno == errno.EAGAIN:
               # the pipe's empty, good
               break
            # else bomb out
            raise SystemExit, "yo2py unexpected IOError:" + str(e)
         except Exception, ee:
            raise SystemExit, "yo2py unexpected Exception:" + str(ee)
         return True

   def set_cursor_busy(self,state):
      if state:
         self.window.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.WATCH))
      else:
         self.window.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.LEFT_PTR))
         

if len(sys.argv) != 2:
   print 'Usage: cubeview.py path/to/cubeview.glade'
   raise SystemExit

cubeview_glade = str(sys.argv[1])
top=cubeview(cubeview_glade)
